utils/stream_compressor.cc

```diff
@@ -68,12 +68,44 @@ static void check_zstd(size_t ret, const char* text) {
     }
 }
 
+// IMPORTANT: with this constant, we are hardcoding a max window size in the decoder.
+//
+// We do this to make the allocator usage by ZSTD more predictable.
+// (If you let zstd alloc and free memory on its own,
+// it has some "helpful" heuristics which sometimes reallocate the buffers
+// even if the context is reused, and we want to avoid that.
+// See https://github.com/facebook/zstd/commit/3d523c741be041f17c28e43b89ab6dfcaee281d2)
+//
+// But the hardcoding means that the decoder won't be able to handle frames with a window
+// size bigger than this. (The decoding will return an error that there's not enough memory).
+//
+// That means that the window size becomes a part of the protocol:
+// the compressor mustn't send messages with a greater window size.
+//
+// If you wish to enlarge the window size for whatever reason,
+// you have to ensure that the new version of the compressor isn't talking
+// to a decompressor with a window size hardcoded to something too small.
+//
+// (In case of RPC compression, you can do this by bumping the COMPRESSOR_NAME
+// in advanced_rpc_compressor.cc when releasing the new Scylla version).
+constexpr size_t ZSTD_WINDOW_SIZE_LOG = 17;
+
```

Comment:
How will that work with pre-patch versions? These won't have 17 hardcoded.

Reply:
> These won't have 17 hardcoded.

Both pre-patch and post-patch versions have compressor-side window size hardcoded.
In this patch, we hardcode the max window size accepted by the decompressor.

Pre-patch versions send messages with hardcoded window size 128 kiB, and accept messages with any window size. Post-patch versions send messages with hardcoded window size 128 kiB, and accept messages with window size smaller than the hardcoded <=128kiB.

So there's no problem. But future patches will have to bump the protocol version if they want to send messages with window size >128 kiB. (Before the patch, they wouldn't have to do that, because before the patch the receiver accepts arbitrarily large window sizes).

Reply:
I see it now, and you did change the 17 in the compressor init to the constant.